[NAOT][android] Application linking breaks with NDK r29 because of duplicated libunwind symbols

### Description

Attempting to link the .NET for Android [NAOT sample](https://github.com/dotnet/android/tree/main/samples/NativeAOT) breaks when [NDK r29](https://developer.android.com/ndk/downloads/index.html#stable-downloads) is installed, with the following error (output broken up into lines to make it easier to read):

```
Android (13989888, +pgo, +bolt, +lto, +mlgo, based on r563880c) clang version 21.0.0 (https://android.googlesource.com/toolchain/llvm-project 5e96669f06077099aa41290cdb4c5e6fa0f59349)
Target: aarch64-unknown-linux-android21
Thread model: posix
InstalledDir: ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin
 "~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld" -EL
 --fix-cortex-a53-843419 -z now -z relro -z max-page-size=16384 --no-rosegment --hash-style=both
 --eh-frame-hdr -m aarch64linux -shared
 -o bin/Release/net10.0-android/android-arm64/native/NativeAOT.so
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21/crtbegin_so.o
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib
 --compress-debug-sections=zlib obj/Release/net10.0-android/android-arm64/native/NativeAOT.o
 --version-script=obj/Release/net10.0-android/android-arm64/native/NativeAOT.exports
 --export-dynamic
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Native.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Globalization.Native.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.IO.Compression.Native.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Security.Cryptography.Native.Android.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbootstrapperdll.o
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libeventpipe-disabled.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libstandalonegc-disabled.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libaotminipal.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlienc.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlidec.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlicommon.a
 --build-id=sha1 --as-needed -e 0x0 -ldl -lz -llog -lm -z relro -z now --eh-frame-hdr -z
 max-page-size=16384 -soname libNativeAOT.so --error-unresolved-symbols --no-undefined
 ~/vc/xamarin/xamarin-android-worktrees/ndk-r29/bin/Debug/lib/packs/Microsoft.Android.Runtime.NativeAOT.36.android-arm64/36.1.99/runtimes/android-arm64/native/libnaot-android.release-static-release.a
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_static.a
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
 obj/Release/net10.0-android/android-arm64/android/jni_init_funcs.arm64-v8a.o
 obj/Release/net10.0-android/android-arm64/android/environment.arm64-v8a.o --discard-all
 --gc-sections -T obj/Release/net10.0-android/android-arm64/native/sections.ld
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/libclang_rt.builtins-aarch64-android.a
 -l:libunwind.a -ldl -lc
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/libclang_rt.builtins-aarch64-android.a
 -l:libunwind.a -ldl
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21/crtend_so.o
ld.lld: error: duplicate symbol: __unw_init_local
>>> defined at libunwind.cpp:44 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:44)
>>>            libunwind.cpp.o:(__unw_init_local) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_init_local+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: libunwind::LocalAddressSpace::sThisAddressSpace
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(libunwind::LocalAddressSpace::sThisAddressSpace) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.bss._ZN9libunwind17LocalAddressSpace17sThisAddressSpaceE+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_get_reg
>>> defined at libunwind.cpp:99 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:99)
>>>            libunwind.cpp.o:(__unw_get_reg) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_get_reg+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_set_reg
>>> defined at libunwind.cpp:114 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:114)
>>>            libunwind.cpp.o:(__unw_set_reg) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_set_reg+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_get_fpreg
>>> defined at libunwind.cpp:146 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:146)
>>>            libunwind.cpp.o:(__unw_get_fpreg) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_get_fpreg+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_set_fpreg
>>> defined at libunwind.cpp:161 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:161)
>>>            libunwind.cpp.o:(__unw_set_fpreg) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_set_fpreg+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_step
>>> defined at libunwind.cpp:197 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:197)
>>>            libunwind.cpp.o:(__unw_step) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_step+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_step_stage2
>>> defined at libunwind.cpp:207 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:207)
>>>            libunwind.cpp.o:(__unw_step_stage2) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_step_stage2+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_get_proc_info
>>> defined at libunwind.cpp:212 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:212)
>>>            libunwind.cpp.o:(__unw_get_proc_info) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_get_proc_info+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_resume
>>> defined at libunwind.cpp:224 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:224)
>>>            libunwind.cpp.o:(__unw_resume) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_resume+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_get_proc_name
>>> defined at libunwind.cpp:238 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:238)
>>>            libunwind.cpp.o:(__unw_get_proc_name) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_get_proc_name+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_is_fpreg
>>> defined at libunwind.cpp:251 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:251)
>>>            libunwind.cpp.o:(__unw_is_fpreg) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_is_fpreg+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_regname
>>> defined at libunwind.cpp:265 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:265)
>>>            libunwind.cpp.o:(__unw_regname) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_regname+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_is_signal_frame
>>> defined at libunwind.cpp:270 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:270)
>>>            libunwind.cpp.o:(__unw_is_signal_frame) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_is_signal_frame+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_iterate_dwarf_unwind_cache
>>> defined at libunwind.cpp:303 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:303)
>>>            libunwind.cpp.o:(__unw_iterate_dwarf_unwind_cache) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_iterate_dwarf_unwind_cache+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_add_dynamic_fde
>>> defined at libunwind.cpp:312 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:312)
>>>            libunwind.cpp.o:(__unw_add_dynamic_fde) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_add_dynamic_fde+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_remove_dynamic_fde
>>> defined at libunwind.cpp:331 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:331)
>>>            libunwind.cpp.o:(__unw_remove_dynamic_fde) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_remove_dynamic_fde+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_add_dynamic_eh_frame_section
>>> defined at libunwind.cpp:336 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:336)
>>>            libunwind.cpp.o:(__unw_add_dynamic_eh_frame_section) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_add_dynamic_eh_frame_section+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: __unw_remove_dynamic_eh_frame_section
>>> defined at libunwind.cpp:358 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:358)
>>>            libunwind.cpp.o:(__unw_remove_dynamic_eh_frame_section) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.text.__unw_remove_dynamic_eh_frame_section+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a

ld.lld: error: duplicate symbol: unw_local_addr_space
>>> defined at libunwind.cpp:38 (/__w/1/s/src/runtime/src/native/external/llvm-libunwind/src/libunwind.cpp:38)
>>>            libunwind.cpp.o:(unw_local_addr_space) in archive ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
>>> defined at libunwind.cpp
>>>            libunwind.cpp.o:(.data.unw_local_addr_space+0x0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64/libunwind.a
```

I hoped that passing `-unwindlib=none` would make linking work, but unfortunately, it causes the following error:

```
Android (13989888, +pgo, +bolt, +lto, +mlgo, based on r563880c) clang version 21.0.0 (https://android.googlesource.com/toolchain/llvm-project 5e96669f06077099aa41290cdb4c5e6fa0f59349)
Target: aarch64-unknown-linux-android21
Thread model: posix
InstalledDir: ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin
 "~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/ld.lld" -EL
 --fix-cortex-a53-843419 -z now -z relro -z max-page-size=16384 --no-rosegment --hash-style=both
 --eh-frame-hdr -m aarch64linux -shared -o
 bin/Release/net10.0-android/android-arm64/native/NativeAOT.so
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21/crtbegin_so.o
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/aarch64
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android
 -L~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib
 --compress-debug-sections=zlib obj/Release/net10.0-android/android-arm64/native/NativeAOT.o
 --version-script=obj/Release/net10.0-android/android-arm64/native/NativeAOT.exports
 --export-dynamic
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Native.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Globalization.Native.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.IO.Compression.Native.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Security.Cryptography.Native.Android.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbootstrapperdll.o
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libeventpipe-disabled.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libstandalonegc-disabled.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libaotminipal.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlienc.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlidec.a
 ~/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlicommon.a
 --build-id=sha1 --as-needed -e 0x0 -ldl -lz -llog -lm -z relro -z now --eh-frame-hdr -z
 max-page-size=16384 -soname libNativeAOT.so --error-unresolved-symbols --no-undefined
 ~/vc/xamarin/xamarin-android-worktrees/ndk-r29/bin/Debug/lib/packs/Microsoft.Android.Runtime.NativeAOT.36.android-arm64/36.1.99/runtimes/android-arm64/native/libnaot-android.release-static-release.a
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_static.a
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
 obj/Release/net10.0-android/android-arm64/android/jni_init_funcs.arm64-v8a.o
 obj/Release/net10.0-android/android-arm64/android/environment.arm64-v8a.o --discard-all
 --gc-sections -T obj/Release/net10.0-android/android-arm64/native/sections.ld
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/libclang_rt.builtins-aarch64-android.a
 -ldl -lc
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/clang/21/lib/linux/libclang_rt.builtins-aarch64-android.a
 -ldl
 ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/lib/aarch64-linux-android/21/crtend_so.o
ld.lld: error: undefined symbol: _Unwind_Resume
>>> referenced by stdexcept:0 (out/lib/device-libcxx-aarch64-ndk-cxx/include/c++/v1/stdexcept:0)
>>>               hash.cpp.o:(std::__ndk1::__throw_overflow_error[abi:ne210000](char const*)) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_static.a
>>> referenced by thread.cpp:0 (out/llvm-project/libcxx/src/thread.cpp:0)
>>>               thread.cpp.o:(std::__ndk1::__thread_local_data()) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_static.a
>>> referenced by call_once.cpp:0 (out/llvm-project/libcxx/src/call_once.cpp:0)
>>>               call_once.cpp.o:(std::__ndk1::__call_once(unsigned long volatile&, void*, void (*)(void*))) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_static.a
>>> referenced 198 more times

ld.lld: error: undefined symbol: _Unwind_RaiseException
>>> referenced by cxa_exception.cpp:295 (out/llvm-project/libcxxabi/src/cxa_exception.cpp:295)
>>>               cxa_exception.cpp.o:(__cxa_throw) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a

ld.lld: error: undefined symbol: _Unwind_SetGR
>>> referenced by cxa_personality.cpp:556 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:556)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
>>> referenced by cxa_personality.cpp:558 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:558)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
>>> referenced by cxa_personality.cpp:556 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:556)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
>>> referenced 1 more times

ld.lld: error: undefined symbol: _Unwind_SetIP
>>> referenced by cxa_personality.cpp:560 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:560)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
>>> referenced by cxa_personality.cpp:560 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:560)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a

ld.lld: error: undefined symbol: _Unwind_GetLanguageSpecificData
>>> referenced by cxa_personality.cpp:626 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:626)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a

ld.lld: error: undefined symbol: _Unwind_GetIP
>>> referenced by cxa_personality.cpp:641 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:641)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a

ld.lld: error: undefined symbol: _Unwind_GetRegionStart
>>> referenced by cxa_personality.cpp:644 (out/llvm-project/libcxxabi/src/cxa_personality.cpp:644)
>>>               cxa_personality.cpp.o:(__gxx_personality_v0) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a

ld.lld: error: undefined symbol: _Unwind_DeleteException
>>> referenced by cxa_exception.cpp:586 (out/llvm-project/libcxxabi/src/cxa_exception.cpp:586)
>>>               cxa_exception.cpp.o:(__cxa_end_catch) in archive ~/android-toolchain/ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a
```

The problem, I think, stems from the fact that `libRuntime.WorkstationGC.a` exports `libunwind` symbols where it should keep them hidden:

```shell
microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native$ nm -C libRuntime.WorkstationGC.a | grep __unw
0000000000000000 T __unw_add_dynamic_eh_frame_section
0000000000000000 T __unw_add_dynamic_fde
0000000000000000 T __unw_get_fpreg
0000000000000000 T __unw_get_proc_info
0000000000000000 T __unw_get_proc_name
0000000000000000 T __unw_get_reg
0000000000000000 T __unw_get_save_loc
0000000000000000 T __unw_init_local
0000000000000000 T __unw_is_fpreg
0000000000000000 T __unw_is_signal_frame
0000000000000000 T __unw_iterate_dwarf_unwind_cache
0000000000000000 T __unw_regname
0000000000000000 T __unw_remove_dynamic_eh_frame_section
0000000000000000 T __unw_remove_dynamic_fde
0000000000000000 T __unw_resume
0000000000000000 T __unw_set_fpreg
0000000000000000 T __unw_set_reg
0000000000000000 T __unw_step
0000000000000000 T __unw_step_stage2
0000000000000000 T __unw_getcontext
```

NDK r29 uses LLVM 21

### Reproduction Steps

1. Check out the https://github.com/dotnet/android/ `dev/grendel/ndk-r29` branch locally (from [this PR](https://github.com/dotnet/android/pull/9926))
2. Go to the checked-out directory and run:
   1. On Linux/macOS: `make prepare all`
   2. On Windows: `build.cmd`
3. After the build completes, go to the `samples/NativeAOT` directory and run
   1. On Linux/macOS: `../../dotnet-local.sh build -c Release -bl`
   2. On Windows: `..\..\dotnet-local.cmd build -c Release -bl`


### Expected behavior

Application should be linked without errors

### Actual behavior

Application linking fails with errors

### Regression?

_No response_

### Known Workarounds

No known workarounds

### Configuration

- .NET version: 10.0.100
- OS: not dependent on the operating system
- Architecture: not dependent on the architecture
- Specific to building a .NET for Android app using NativeAOT with NDK r29


### Other information

_No response_

I was able to repro this on a fresh vm. I have a workaround that localizes these libunwind symbols, but then I get a similar failure:

```
Restore complete (0.2s)
  NativeAOT net10.0-android android-arm64 failed with 6 error(s) (0.3s)
    ld.lld : error duplicate symbol: RhCollect
    ld.lld : error duplicate symbol: RhGetGcTotalMemory
    ld.lld : error duplicate symbol: RhStartNoGCRegion
    ld.lld : error duplicate symbol: RhEndNoGCRegion
    clang : error linker command failed with exit code 1 (use -v to see invocation)
    /home/sven/.nuget/packages/microsoft.dotnet.ilcompiler/10.0.0/build/Microsoft.NETCore.Native.targets(389,5): error MSB3073: The command ""clang" "obj/Release/net10.0-android/android-arm64/native/NativeAOT.o" -o "bin/Release/net10.0-android/android-arm64/native/NativeAOT.so" -Wl,--version-script="obj/Release/net10.0-android/android-arm64/native/NativeAOT.exports" -Wl,--export-dynamic -gz=zlib -fuse-ld=lld /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Native.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Globalization.Native.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.IO.Compression.Native.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libSystem.Security.Cryptography.Native.Android.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbootstrapperdll.o /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libRuntime.WorkstationGC.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libeventpipe-disabled.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libstandalonegc-disabled.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libaotminipal.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlienc.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlidec.a /home/sven/.nuget/packages/microsoft.netcore.app.runtime.nativeaot.android-arm64/10.0.0/runtimes/android-arm64/native/libbrotlicommon.a --target=aarch64-linux-android21 -g -Wl,--build-id=sha1 -Wl,--as-needed -Wl,-e,0x0 -pthread -ldl -lz -llog -lm -shared -Wl,-z,relro -Wl,-z,now -Wl,--eh-frame-hdr -Wl,-z,max-page-size=16384 "-Wl,-soname,libNativeAOT.so" -Wl,--error-unresolved-symbols -Wl,--no-undefined "/home/sven/android/bin/Debug/lib/packs/Microsoft.Android.Runtime.NativeAOT.36.android-arm64/36.1.99/runtimes/android-arm64/native/libnaot-android.release-static-release.a" "/home/sven/maui-android-native/android-sdk/ndk/29.0.14206865/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++_static.a" "/home/sven/maui-android-native/android-sdk/ndk/29.0.14206865/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib/aarch64-linux-android/libc++abi.a" obj/Release/net10.0-android/android-arm64/android/jni_init_funcs.arm64-v8a.o obj/Release/net10.0-android/android-arm64/android/environment.arm64-v8a.o -Wl,--discard-all -Wl,--gc-sections -Wl,-T,"obj/Release/net10.0-android/android-arm64/native/sections.ld"" exited with code 1.
  NativeAOT net10.0-android failed with 1 warning(s) (0.2s) → bin/Release/net10.0-android/NativeAOT.dll
    /home/sven/android/bin/Debug/lib/packs/Microsoft.Android.Sdk.Linux/36.1.99/tools/Xamarin.Android.Common.targets(570,3): warning XA1040: The NativeAOT runtime on Android is an experimental feature and not yet suitable for production use. File issues at: https://github.com/dotnet/android/issues

Build failed with 6 error(s) and 1 warning(s) in 1.3s
```

which seems to be caused by the compiler emitting some symbol definitions for these into the application object file (and these conflict with the definitions in libRuntime.WorkStationGC.a):

```
> nm obj/Release/net10.0-android/android-arm64/native/NativeAOT.o | grep -E "(RhCollect|RhGetGcTotalMemory|RhStartNoGCRegion|RhEndNoGCRegion)"
00000000000e5300 T RhCollect
00000000000e5460 T RhEndNoGCRegion
00000000000e5370 T RhGetGcTotalMemory
00000000000e53d0 T RhStartNoGCRegion
00000000000e5300 T S_P_CoreLib_System_Runtime_InternalCalls__RhCollect
00000000000e5460 T S_P_CoreLib_System_Runtime_InternalCalls__RhEndNoGCRegion
00000000000e5370 T S_P_CoreLib_System_Runtime_InternalCalls__RhGetGcTotalMemory
00000000000e53d0 T S_P_CoreLib_System_Runtime_InternalCalls__RhStartNoGCRegion
000000000001336f R _lsda0S_P_CoreLib_System_Runtime_InternalCalls__RhStartNoGCRegion
```

Is that expected @MichalStrehovsky? It looks like a bug to me.

--- a/src/coreclr/nativeaot/Runtime/CMakeLists.txt
+++ b/src/coreclr/nativeaot/Runtime/CMakeLists.txt
@@ -149,8 +149,12 @@ else()
     include_directories($ENV{EMSCRIPTEN/system/lib/libcxxabi/include})
   endif()
 
-  # Disable building _Unwind_XXX style APIs of libunwind, since we don't use them.
-  add_definitions(-D_LIBUNWIND_DISABLE_ZERO_COST_APIS=1)
+  # Guard out the public unw_* C API, C++ exception dispatch entry points,
+  # and the LocalAddressSpace singleton from llvm-libunwind. NativeAOT does
+  # not use these - it uses the internal C++ classes directly. Omitting them
+  # avoids exporting symbols that conflict with platform libunwind (e.g.
+  # Android NDK r29).
+  add_definitions(-D_LIBUNWIND_NATIVEAOT=1)
 
   # Compile unwinding only for the current compilation target architecture
   add_definitions(-D_LIBUNWIND_IS_NATIVE_ONLY)

--- a/src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp
+++ b/src/coreclr/nativeaot/Runtime/unix/UnixNativeCodeManager.cpp
@@ -56,7 +56,7 @@ UnixNativeCodeManager::UnixNativeCodeManager(TADDR moduleBase,
       m_pClasslibFunctions(pClasslibFunctions), m_nClasslibFunctions(nClasslibFunctions)
 {
     // Cache the location of unwind sections
-    libunwind::LocalAddressSpace::sThisAddressSpace.findUnwindSections(
+    UnwindHelpers::FindUnwindSections(
         (uintptr_t)pvManagedCodeStartRange, m_UnwindInfoSections);
 }
 

--- a/src/coreclr/nativeaot/Runtime/unix/UnwindHelpers.cpp
+++ b/src/coreclr/nativeaot/Runtime/unix/UnwindHelpers.cpp
@@ -1396,6 +1396,11 @@ bool UnwindHelpers::GetUnwindProcInfo(PCODE pc, UnwindInfoSections &uwInfoSectio
     return true;
 }
 
+bool UnwindHelpers::FindUnwindSections(uintptr_t pc, UnwindInfoSections &sections)
+{
+    return _addressSpace.findUnwindSections(pc, sections);
+}
+
 #if defined(TARGET_APPLE)
 // Apple considers _dyld_find_unwind_sections to be private API that cannot be used
 // by apps submitted to App Store and TestFlight, both for iOS-like and macOS platforms.

--- a/src/coreclr/nativeaot/Runtime/unix/UnwindHelpers.h
+++ b/src/coreclr/nativeaot/Runtime/unix/UnwindHelpers.h
@@ -15,4 +15,5 @@ class UnwindHelpers
 public:
     static bool StepFrame(REGDISPLAY *regs, unw_word_t start_ip, uint32_t format, unw_word_t unwind_info);
     static bool GetUnwindProcInfo(PCODE ip, libunwind::UnwindInfoSections &uwInfoSections, unw_proc_info_t *procInfo);
+    static bool FindUnwindSections(uintptr_t pc, libunwind::UnwindInfoSections &sections);
 };

--- a/src/native/external/llvm-libunwind-version.txt
+++ b/src/native/external/llvm-libunwind-version.txt
@@ -3,3 +3,4 @@ https://github.com/llvm/llvm-project/releases/tag/llvmorg-22.1.1
 
 Apply https://github.com/dotnet/runtime/commit/35b7d59fa1075ab0fefb921393409806a821d8ed
 Apply https://github.com/dotnet/runtime/commit/be5f98fb6702704afbaf705dce0b54d55479c6f1
+Apply https://github.com/dotnet/runtime/commit/fde4a5bd965b128318e17a3b8ba1dc0fcf8df4a0

--- a/src/native/external/llvm-libunwind/src/Unwind-EHABI.cpp
+++ b/src/native/external/llvm-libunwind/src/Unwind-EHABI.cpp
@@ -443,6 +443,12 @@ _Unwind_VRS_Interpret(_Unwind_Context *context, const uint32_t *data,
   return _URC_CONTINUE_UNWIND;
 }
 
+// The C++ exception dispatch functions below depend on __unw_step and other
+// public API functions from libunwind.cpp which are guarded out when
+// _LIBUNWIND_NATIVEAOT is defined. NativeAOT does not use these - it only
+// uses _Unwind_VRS_Interpret above.
+#if !defined(_LIBUNWIND_NATIVEAOT)
+
 extern "C" _LIBUNWIND_EXPORT _Unwind_Reason_Code
 __aeabi_unwind_cpp_pr0(_Unwind_State state, _Unwind_Control_Block *ucbp,
                        _Unwind_Context *context) {
@@ -1209,5 +1215,7 @@ __gnu_unwind_frame(_Unwind_Exception *exception_object,
     return _URC_FAILURE;
   }
 }
+
+#endif // !defined(_LIBUNWIND_NATIVEAOT)
 
 #endif  // defined(_LIBUNWIND_ARM_EHABI)

--- a/src/native/external/llvm-libunwind/src/libunwind.cpp
+++ b/src/native/external/llvm-libunwind/src/libunwind.cpp
@@ -26,6 +26,8 @@
 #include <sanitizer/asan_interface.h>
 #endif
 
+#if !defined(_LIBUNWIND_NATIVEAOT)
+
 #if !defined(__USING_SJLJ_EXCEPTIONS__) && !defined(__wasm__)
 #include "AddressSpace.hpp"
 #include "UnwindCursor.hpp"
@@ -546,6 +548,8 @@ int __unw_remove_find_dynamic_unwind_sections(
 }
 
 #endif // __APPLE__
+
+#endif // !defined(_LIBUNWIND_NATIVEAOT)
 
 // Add logging hooks in Debug builds only
 #ifndef NDEBUG

